### ISSUE TITLE
Container polymorphism

### ISSUE THREAD

**Issue by byorgey** [disco-lang/disco#127]
It will considerably simplify things involving containers if we

* Add "container variables" in addition to type variables
* Add equality and "subcontainer" constraints on container variables
* Extend the solver to solve for these.  (It should be easy enough since they form a simple lattice, with none of the complication of sorts etc.)

We probably don't want to expose container polymorphism in the disco type system itself.  So various container functions will have to be primitives since we can't really express their types.

**Comment by byorgey** [comment]
The subtyping here is perplexing.  On the one hand it seems that a hierarchy like List < Multiset < Set makes sense, going from more free to less free.  There are appropriate monoid homomorphisms along each of these.  On the other hand there is also a sensible monoid homomorphism Set -> Multiset.

I am coming around to the position that there shouldn't be any container subtyping at all; there should instead be explicit functions `list`, `set`, `multiset` which will convert *to* the given container type *from* any other container type.  We have the following Galois connections (actually just section/retraction pairs):
```
      --- forget order -->          --- forget cardinalities -->
List                       Multiset                              Set
      <-- sorted order ---          <-- assign card. 1 to all --
```
Going left and then right in this diagram is the identity.  Going right and then left is not, in general, the identity, but represents some sort of (idempotent) canonicalization: `List -> Multiset -> List` is `sort`; `Multiset -> Set -> Multiset` drops duplicates; `List -> Set -> List` both sorts and drops duplicates.

So if a function `f` expects a set, you can't call it with a list, `f [1,2,3]`, but you could instead say `f (set [1,2,3])`.  It also won't be possible to write `{ x | x in [1 .. 10], even x }` because that would mix a set and a list; one would have to write `{ x | x in set [1..10], even x }` (actually in this case of course one could just write `x in {1 .. 10}`).

For simplicity I suggest the following protocol for set/multiset literals:
```
{ 1, 2, 3 # 5 }   -- definitely a multiset
{ 1, 2, 3 }   -- definitely a set
multiset {1, 2, 3}  -- you could write this if you explicitly want a multiset
{ 1, 2, 3 # 1 }  -- or this
```
In other words, whether `{ ... }` is a set or multiset is determined *syntactically* by the presence or absence of `#` notation.  One could imagine something slightly more complex where set literals with no hashes will infer as sets but check at either set or multiset type, but it seems simpler just to go with the syntactic distinction.

Note that `#` should be allowed in list literals too.

**Comment by byorgey** [comment]
Not sure what I was thinking before, there isn't a monoid homomorphism Set -> Bag.  At least the obvious one that assigns cardinality 1 to each element isn't a homomorphism.

Also note these are RINGAD morphisms, not just monoid.  See Gibbons.

**Comment by byorgey** [comment]
We have container polymorphism now, except no container *subtyping* yet.  Closing this in favor of more specific tickets.
